db/migrate/20190517093412_create_firmware_constraints.rb

```diff
@@ -0,0 +1,9 @@
+class CreateFirmwareConstraints < ActiveRecord::Migration[5.0]
+  def change
+    create_table :firmware_constraints do |t|
```

Comment:
I'm not sure I understand the name of this table. How does this represent a constraint? It just looks like some shared details.

Reply:
@miha-plesko would it make sense to embed this into the firmware record then?  Why split it out into its own model

Reply:
Agree, this feels like attributes of firmware that should be indexed.

Reply:
Correct me if I'm wrong, but you're suggesting to add those two fields on FirmwareBinary?

```
FirmwareBinary.compatible_manufacturer
FirmwareBinary.compatible_model
```

But usually the same firmware binary can be applied on more than one kind of physical server. How would we handle this then? Should we just duplicate the firmware binary itself?

Reply:
Ah I see. Maybe just change the name to something more descriptive ... like `firmware_targets`? "Constraints" makes me feel like we're implementing a version constraint system of some kind.

What do you think @agrare ?

Reply:
This also feels like an attribute of the physical server, but I guess that relationship isn't being modeled and instead you'll match up the manufacturer and model when you need to do the update, right?

Reply:
Nice, if @agrare also agrees I'll rename like you suggest. Was in search for better name myself but couldn't think of one

Reply:
Ah okay I thought you were modelling this the other way around, that one target platform could use multiple firmwares not one firmware could be used by multiple target platforms.

`:firmware_targets` makes sense then :+1:

Reply:
FWIW @agrare I think it's a many-to-many relationship (judging by the join table :wink: )

Reply:
@agrare it's both ways - that's why we're using many-to-many.

- a firmware can be applied to multiple platforms (like `Dell ZX01`, `Dell ZX8` etc)
- a platform can have many firmwares applied (like `nic-firmware-upodate-5.2.3.exe`, `bios-update-3.1.exe` etc)

@carbonin yeah both `manufacturer` and `model` are attribute on physical server. In theory we could extend inventory parser to link appropriate `firmware_target` as well, but that would bring redundancy. Unless we remove those two attributes from server, but how would you then update a server... It gets complicated

Reply:
Yeah, I was thinking about that too. Since neither is actually dependent on each other you probably need to collect them separately. (e.g. you can have firmware that targets a server that isn't in inventory yet).

You could probably have a separate many-to-many relationship between physical servers and firmware targets, but I don't think that really gains you anything over joining using the manufacturer and model ...